lib/Sema/TypeCheckExpr.cpp

```diff
@@ -522,8 +522,17 @@ static Expr *foldSequence(TypeChecker &TC, DeclContext *DC,
       // If we've drained the entire sequence, we're done.
       if (S.empty()) return LHS;
 
-      // Otherwise, start all over with our new LHS.
-      return foldSequence(TC, DC, LHS, S, precedenceBound);
+      // Otherwise, we have to check that this next operator actually
+      // associates.
+      if ((op2 = getNextOperator()) && op2.precedence) {
+        associativity =
+            TC.Context.associateInfixOperators(op1.precedence, op2.precedence);
+        RHS = S[1];
+      }
+
+      // If so, start all over with our new LHS.
+      if (associativity != Associativity::None)
+        return foldSequence(TC, DC, LHS, S, precedenceBound);
```

Comment:
Hmm, this doesn't seem like the right fix.  Could you try changing `PrecedenceBound::shouldConsider` so that it returns true if the associativity is not right-associative rather than only if it's left-associative?

Reply:
You're right, that's much cleaner. Thanks!